### ISSUE TITLE
Loading too many tiles after #1014

### ISSUE THREAD

**Issue by iangilman** [openseadragon/openseadragon#1020]
After landing #1014 @johnsusi and I have noticed that it's now loading more tiles than it needs. Beforehand (in my experience) it loads the highest level that covers the entire image with a single tile and then progresses upward from there. Now it does that but also progresses downward, loading every single level down to level 0. In @johnsusi's case he had `immediateRender` on, so it's even more dramatic that it's loading everything (instead of just the top level it needed).

Obviously we don't want to be loading these unnecessary tiles. One approach would be to roll back #1014, but I think it's actually the right step forward, and it's just uncovered an existing flaw in our logic, so I propose we leave the patch and seek to fix this new issue.

One theory: When we go through looking for the next tile to load perhaps we don't have a "don't bother" threshold. Now with #1014 we keep looking for next tiles to load more aggressively, so we need to institute that threshold. It should be based on the knowledge that the tiles needed to cover the scene are already in process somehow.

Note that if a high level tile fails to load, we should then fall back to loading a tile below it to cover that space. Perhaps our current code treats "this tile is currently being loaded" as a load failure.

@johnsusi says he'll dig into it, and perhaps @CaptainCodeman can be of help as well.

**Comment by CaptainCodeman** [comment]
Yeah, I noticed that today ... I thought I just had the immediateRender set wrong at first but I guess not.

I would guess that there is something else that the tile-loaded code does or it's simply a race condition - when it's working out whether a tile should be loaded, it's not taking into account the ones that are in the process of loading.

**Comment by CaptainCodeman** [comment]
I've checked again an as far as I can tell it appears to be doing the right thing (or what I expect at least) - if I set `immediateRender: true` then it doesn't load the lower-level (zoomed out) versions first or any additional tiles after.

Maybe the size combination doesn't show it up for the images I've looked at - is there a particular test version you're using so I can confirm?

**Comment by johnsusi** [comment]
Time seems to be a factor here. If the first tiles load fast enough then the it doesn't load the lower levels but if I throttle the network connection then it loads down to level 0 for me.

**Comment by CaptainCodeman** [comment]
What example is that with? (or what options / settings)

**Comment by iangilman** [comment]
I can get it to happen with http://localhost:8000/test/demo/basic.html ...without throttling the bandwith it still loads down to level 5, which is lower than it should. If I use Chrome's Network Inspector to throttle it to "Regular 3G" it goes all the way down to level 0.

**Comment by iangilman** [comment]
Yeah, something like that. Thank you for digging into it!

**Comment by rdlester** [comment]
If I understand @CaptainCodeman correctly, it sounds like the issue is that OpenSeadragon has a bit of error handling that attempts to ensure there is always at least one image on screen. This prevents errors in fetching all the Tiles that OpenSeadragon would normally pick from leaving the screen blank if there's at least one Tile in the entire image it can download. However, when determining whether it has successfully fetched any Tiles, it only checks `Tile.loaded`, not `Tile.loading`, meaning it interprets the case where Tiles download successfully but slowly as a complete failure to download any Tiles and starts prematurely scheduling download jobs for additional tiles.

This wasn't exposed before #1014 b/c it was impossible for OpenSeadragon to reach a state where download jobs were scheduled for all Tiles OpenSeadragon normally downloads.

I'd guess there are two things to fix here:

- This error handling logic should be checking `Tile.exists` and entirely ignore loading state.

- `TiledImage._needsDraw` should only get set after firing a loading job for a Tile if there are still more Tiles to download. If it's dispatched a job for the last Tile, that flag shouldn't get flipped. After this, drawing should only restart when a download job succeeds or fails. This is trivial after #173 is completed, as all required download jobs will be guaranteed to be scheduled after a single run of the render loop, and `_needsDraw` will never need to get flipped for the purpose of loading more Tiles except maybe on download failure, which is already handled.

**Comment by rdlester** [comment]
Actually, the root problem is more insidious.

When determining Tiles to download, `updateViewport` always loops from the highest resolution level it makes sense to display all the way down to the bottom-most level. It never breaks early.

While considering levels, `updateViewport` tracks whether it has drawn any Tiles as of yet in `haveDrawn`. When considering an individual level, it only sets `drawLevel` immediately to true when `!haveDrawn` and some particular pixel ratio is satisfied that I don't really understand. After that, `drawLevel` is false. Presumably the highest level always gets `drawLevel` set to true and the rest to false. 

However, when considering an individual Tile, `drawLevel` can be overridden and flipped to true if `haveDrawn` and the Tile is not already covered. Since no coverage is set until a Tile is loaded, this means the first run will consider all Tiles.

Finally, if the Tile is neither loaded nor loading, and `drawLevel` is true (either by initial set or override), it will compare the Tile with the current "most downloadable" Tile. Presumably, OpenSeadragon ensures that too low a level is never drawn by comparison against `Tile.visibility`. However, if all the expected Tiles are downloaded or downloading, there will be no "most downloadable" Tile upon reaching overly low levels. Usually, the coverage check is enough to prevent these overly low levels from getting downloaded; however, if all Tiles are still downloading, no coverage will be set, and these overly low Tiles will become the best by merit of there being no other Tiles to download.

**Comment by rdlester** [comment]
I don't see an easy way to fix this particular issue. However, as mentioned above, this should no longer cause problems after #173.

**Comment by rdlester** [comment]
Actually, it's still a problem for #173, as #173 needs to know exactly what set of Tiles is the correct set to download (excluding tiles at too low a level), and this issue means it's not clear how to determine that.

**Comment by CaptainCodeman** [comment]
Yeah, there isn't a simple solution, at least not one that I could find.

I tried going back to older versions to see if I could find where it was introduced and as far back as I went, it seemed like it always worked that way. So maybe it needs a more fundamental re-working of some pieces to achieve the ideal loading process.

The faster first-load is more important for my needs but maybe for now it should be put behind a config option so people can chose which behavior they want?

**Comment by iangilman** [comment]
@rdlester Well, it seems like #173 is a good step in the right direction, and a good opportunity to clean up a bunch of these issues. Are you up for working on it?

Here's how I see the new logic working:

Basically we're talking about breaking apart the "what tiles should I load" calculation from the "draw the tiles I have" process.

When the tiled image first appears in the viewer, we go through and generate a prioritized list of tiles it needs to load. It considers all levels that "make sense" for its current position: from the lowest (i.e. lowest resolution) level that fits the entire image into a single tile to the highest level that doesn't provide higher resolution than we need at the moment. The tiles are set to load in the order we want them, starting with that single tile level, and then the next level, and so on (so we never experience holes in the image, and the image resolves smoothly). On multi-tile levels, we prioritize tiles that are closest to the center of the viewport. During this process we assume that we'll be able to get all tiles successfully, so we don't consider fallbacks at lower levels. 

This queue is then stored and we work off of it to load the tiles, respecting the imageLoaderLimit. As each new tile appears we do a redraw.

When anything changes in the geometry of the scene (zoom/pan/image movement), we throw out our current queue and generate a fresh queue again. Same if one of the tiles comes back with an error. With errors noted we may need to go into lower levels to ensure coverage.

Sound about right?

**Comment by rdlester** [comment]
Yeah, sounds correct to me, assume by "lowest level that fits the entire image in a single tile" you mean "highest level whose tiles in image-space are still larger than the viewport (in image-space)".

Of course, going all the way down to the very lowest level would avoid running into an error where all the tiles currently being considered fail to download but those at lower levels still load fine. That seems absurdly rare, though, so probably not something to worry about.

I can take a stab at this, sure.

**Comment by iangilman** [comment]
> Yeah, sounds correct to me, assume by "lowest level that fits the entire image in a single tile" you mean "highest level whose tiles in image-space are still larger than the viewport (in image-space)".

Well, I guess actually I mean *highest* level that fits the entire image into a single tile. In other words, maybe level 7 just has tile 0_0 and no other tiles, but level 8 has 4 tiles (0_0, 0_1, 1_0, 1_1); in that case you would start with level 7, since it can give you all of the information you need to cover the entire image in a single call to the server. You ignore level 0 through 6, since even though they are smaller, they still represent single round trips and therefore aren't really any more efficient.

Incidentally, levels 0 through 6 (in that scenario... of course the level numbers may be different for different images) are important if we ever get around to implementing the "collection optimization" where the tiny levels of many images are all grouped together into a single tile to support rapid loading of large numbers of images. Meanwhile, we don't need to load them, unless we get an error on one of the higher levels...

> Of course, going all the way down to the very lowest level would avoid running into an error where all the tiles currently being considered fail to download but those at lower levels still load fine. That seems absurdly rare, though, so probably not something to worry about.

My thought is that since this scenario is so rare we shouldn't load those levels ahead of time, but we can still have code that goes back and loads them if it turns out that some of the higher levels do have errors. 

> I can take a stab at this, sure.

Awesome, I'm looking forward to it!

**Comment by johnsusi** [comment]
This sounds like the correct way forward. 

However I would actually want the highest level that covers the entire image to be loaded, but not first. This helps when an image is loaded and only a part of the image is visible to the user but then the user zooms out. If higher levels are not loaded then the user sees the background instead of the (lores) image.

So my preferred load order would be:
[ 
  tile-that-spans-entire-view, 
  ... , 
  ... tiles-at-current-resolution, 
  tile-that-spans-entire-image, 
  ..., 
  level-above-tile-that-spans-entire-view
]

Obviously this will not help if the user zooms out to fast but it is good enough for our usecase.

**Comment by iangilman** [comment]
I see... I hadn't thought of the "already zoomed in to a spot" scenario. Even there, for the general case, I think I'd rather get the tile that spans the entire image first, in case of zooming out. Never showing a partial image is a big point in the OSD philosophy. 

I feel less strongly about how one gets from the image span tile to the current resolution... traditionally we go up one level at a time, but maybe it's worth revisiting that. Maybe we go straight from the tile spanning image to the view spanning image? And then from there maybe we could go up every other level or some other faster approach? It just needs to be gradual enough that you don't see dramatic tile seams (where there's one spot on the image that's super sharp and another part that's super blurry). Going straight from the tile spanning image to the view spanning image won't introduce seams, though, since they both cover the entire viewport.

**Comment by johnsusi** [comment]
Perhaps the selection algorithm can be implemented using the strategy pattern so clients can provide their own behaviour if they do not like the default.

**Comment by iangilman** [comment]
Indeed, good idea.

**Comment by iangilman** [comment]
@rdlester Have you had a chance to dig into this yet? How's it looking?

**Comment by rdlester** [comment]
I've been busy the past two weeks but will hopefully have some free time again in a few days. I'll update then regardless.

**Comment by iangilman** [comment]
Great, sounds good :)

**Comment by iangilman** [comment]
@rdlester Any progress?

**Comment by rdlester** [comment]
Not yet, got busy with another project. Will take another look at this over the holidays.

**Comment by iangilman** [comment]
Cool :)

**Comment by iangilman** [comment]
@rdlester Any news?

**Comment by iangilman** [comment]
Well, this is one of the last issues before we release 2.3.0. I'm tempted to roll back #1014 for the release and then we can continue working on this afterwards.

**Comment by rdlester** [comment]
Sorry for the silence, my personal computer died a couple weeks ago and I've only just obtained a new one, which prevented me from working on this. Ramping back up on this now. But, I'd recommend going forward with the release without this, as I can't make any guarantees about timeline.

**Comment by iangilman** [comment]
Oh no! Sorry to hear that. Glad you're up and running again. Keep us posted :)

**Comment by CaptainCodeman** [comment]
Yeah, death of a PC is never nice.

Maybe I'm not understanding the issue but the loading seems to be working as expected for me.

I'm seeing it load images in parallel for the initial view but stop when it's reached the appropriate tile level and it not load more beyond that. But if I do zoom in further then the additional higher resolution tiles are loaded.

I'll pull the latest changes and do some more testing, I wonder if it's some combination of options I have set? If so that might give hints as to how to fix it.

**Comment by iangilman** [comment]
@CaptainCodeman That certainly sounds like what we're hoping for it to do... thanks for taking a look; I look forward to hearing what you find!

Note that @johnsusi was able to repro more easily with `immediateRender` on.

**Comment by CaptainCodeman** [comment]
Ah, I may have misunderstanding the issue. Here's what I see when I set the demo/basic.html page to load tileSources: "http://openseadragon.github.io/example-images/duomo/duomo.dzi" (and debugMode: true to help visualize things).

It starts at level 7 and then stops at level 10 (the tiles are then about 235px each).
No other tiles have been loaded at that point (looking at the network trace).
If I zoom in then level 11 loads ... then 12 when zoomed in further and so on up to level 14 (maximum zoom).

What you're talking about is it loading the smaller full image tiles (i.e. 6, 5, 4 ...). I only see those if I throttle the network connection right down and I realize I set the minLevel in my app to prevent that ever happening.

It's strange that it doesn't start with the smallest size though if it's meant to give a "progressive jpeg" rendering effect of it coming into focus. It does seem rather pointless to ever load any image that is smaller than the largest single tile that represents the complete image (once you factor in jpeg overhead, http traffic etc...)

**Comment by johnsusi** [comment]
I worked around this issue by turning off immedateRender and restricted the zoom-level so now it only loads level 10 and above (below?), nothing smaller.

**Comment by iangilman** [comment]
@CaptainCodeman I could see how if your network connection is fast enough it would never get around to loading the lower levels, as it's a timing issue... it loads the lower levels in between the time it launches the request for the higher levels and the time when it receives the higher levels. As far as I can tell, this is because it counts "is loading a tile" as "failed to load the tile" as far as its decision whether to launch another load request.

And yes, I agree in normal circumstances you shouldn't ever load an image that is smaller than the largest single tile that represents the complete image. The one exception is if that larger tile fails to load; then you'll want to fall back on the next smaller tile.

**Comment by harukizaemon** [comment]
I'm having what appears to be a similar issue.

I have an image (tiled at 256 pixels) that is physically larger (6300x3000) than the viewport (512x512). There are 14 levels (0-13). OSD begins by loading `10/...` then mysteriously loads in `11/...`.

Is this related? Is there a workaround?

**Comment by iangilman** [comment]
@harukizaemon That may be perfectly normal... if level 10 is the largest level that's contained in a single tile, OSD would open that first, but then continue onto level 11 for higher resolution. 

What we're concerned about here is if it then starts going back down the levels and opening 9, 8, etc.

**Comment by harukizaemon** [comment]
@iangilman Ahh. In this case it was four tiles, yes and then it began loading the next higher level of resolution. So long as that's normal, all good. Thank you!

**Comment by iangilman** [comment]
Okay, I think I've hit upon a solution for this issue. Please take a look at https://github.com/openseadragon/openseadragon/pull/1193 and see what you think. I'd appreciate it if someone could try out the patch to make sure it's working as expected.

**Comment by VipSaran** [comment]
👍

**Comment by susanmonnelly** [comment]
@iangilman I pulled  down your branch and tried it out and I'm not seeing tiles load when I zoom out  ( sorry I can't post a demo since we are behind a firewall) .  I did take screen shots that I can post . 
This is the list of images loaded on the initial load and you can see when I zoom out there are two tiles that load at a later time frame and the image is displayed in the viewer:

![working](https://cloud.githubusercontent.com/assets/12297503/26121570/bb87cc30-3a42-11e7-8b16-9e05e8acca06.png)

When I use your branch you can see there are fewer tiles loaded which is the behavior we want.  I have immediateRender set to true.   Then when I zoom out, the other tiles do not load and you can see the viewer is not displaying the image.  I'm loading the same image in both examples and have zoomed out completely in both.

![notworking](https://cloud.githubusercontent.com/assets/12297503/26121680/12b5fa5e-3a43-11e7-8bbb-ff869d2b9657.png)

I assume your tests are working when zooming out and trying to load more tiles?

Susan

**Comment by iangilman** [comment]
@susanmonnelly Yeah, zooming out is working fine for me on this patch. Here's my test code (meant for the test/demo/basic.html file in this repository):

```
var duomo = {
    Image: {
        xmlns: "http://schemas.microsoft.com/deepzoom/2008",
        Url: "http://openseadragon.github.io/example-images/duomo/duomo_files/",
        Format: "jpg",
        Overlap: "2",
        TileSize: "256",
        Size: {
            Width:  "13920",
            Height: "10200"
        }
    }
};

var viewer = OpenSeadragon({
    debugMode: true,
    immediateRender: true,
    id: "contentDiv",
    minZoomLevel: 0.01,
    prefixUrl: "../../build/openseadragon/images/",
    tileSources: duomo
});
```

I wonder what's different between your project and mine? How big is your image? How responsive is your server? Any other thoughts? I'd love to get to the bottom of this. Thank you for testing it!

**Comment by susanmonnelly** [comment]
Hi @iangilman 

I tried a smaller image today and still had the same problem.  Here are  the contents of .dzi file I pass to tileSources:

```
<Image Format="jpeg" Overlap="1" TileSize="254">
   <Size Height="950" Width="1639"/>
</Image>
```

This is my viewer creation:
```
    // create a viewer with the OpenSeadragon() function.
    _viewer = OpenSeadragon({
      id: _data.nbViewerID,
      toolbar: "NBViewerOSDToolbar",
      constrainDuringPan: true,
      visibilityRatio: 1,
      showNavigator:  true,
      navigatorAutoFade: false,
      showRotationControl: _data.nbViewerRotationControl,
      // change the size of the navigator window
      navigatorPosition: "ABSOLUTE",
      navigatorAutoResize: false,
      navigatorRotate: false,
      navigatorHeight: "150px",
      navigatorWidth: "150px",
      //debugMode: true,
      immediateRender: true,
      // adds precision to the zoom click
      zoomPerClick: 1.3,
      // adjusts the floating at the end of a zoom
      animationTime: 0.5,
      prefixUrl: _data.nbViewerImagePath,
      tileSources: _data.nbViewerSettings.response.params.dzi_path,
      // Settings remove the zoom functions from the mouse.
      gestureSettingsMouse: { scrollToZoom: false, dblClickToZoom: false, clickToZoom: false },
    });

```

When I change to  `immediateRender:false` it of course works with branch.

We worked on performance so the files in dzi directory are all created by the time I make the ajax request .  I would say the server is pretty responsive.   The strange thing is that zooming in works great and that brings in many more tiles, but zooming out which should loads fewer  tiles, does not work.  Panning works well too.

  Another odd thing is that I don't see anything in the browsers console that the viewer is even trying to load any tiles  when I zoom out.

I just looked at this again with immediteRender=false and when I zoom out, there are no more tiles loaded in so it must be using one of those tiles that was pulled in initially that's in the 0_0 range.  I have one  for every zoom level from 1 to 12. 

Does any of this help? 

Susan

**Comment by iangilman** [comment]
All good info, but I still can't repro it. I've tried using your settings, and both the duomo image (which is high res) and the testpattern.dzi (in this repository) which is low res. 

One question: with your settings, I don't see how to zoom out very far. You should be able to click on the zoom out button once and then it stops you from going further than that. Is that the same for you? What if you add `minZoomLevel: 0.01`? 

What browser are you on? Looks like Firefox. What OS? I've tried on Mac Firefox and Chrome. Have you tried on another browser? 

Would it be possible for you to share the actual tiles for that smaller image? Or could you see if you could get it to repro with one of the test images in this repository? 

Could you try using `debugMode: true` to see if there is anything you notice? 

I've pushed a little console.log to this branch that spits out the level, x, y of every tile it loads. Perhaps it'll be easier to see what's happening (or not happening) with that. 

Thank you for helping debug this!

**Comment by susanmonnelly** [comment]
@iangilman 

I pulled in your latest change but it looks like there were several updates although that could have been your test updates.

While testing this with the console.log statements in place, the zoomed out tiles began to load, even without  changing the `minZoomLevel` .  That could indicate a timing problem at our end I suppose?   I also noticed that when the upper level tiles loaded, it was not a smooth transition.   There was clear indication that there was a delay in loading in the new tiles.

When I get back from vacation on Tuesday, I will package up the tiles in a repo and try sharing it with you.   

I'm using FF on a MAC. HAven't tried other browsers yet.

I'll get back to you on Tues.  Sorry for the delay

SUsan

**Comment by iangilman** [comment]
@susanmonnelly Thank you for the updates; no worries on the timing... I look forward to playing with the repo on Tuesday :)

And yes, I've been merging the latest master into that branch, so those are probably the additional changes you saw.

**Comment by iangilman** [comment]
@susanmonnelly have you had a chance to get that repo together?
